### PR TITLE
Save only top k validation models based on accuracy

### PR DESCRIPTION
**Pull request by jack-willturner**

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -59,11 +59,11 @@ def get_base_argument_parser():
     # logging
     parser.add_argument("--experiment_name", type=str, default="dev")
     parser.add_argument("--logs_path", type=str, default="log")
+    parser.add_argument("--save_k_top_val_models", type=int, default=1)
 
     parser.add_argument("--filepath_to_arguments_json_config", type=str, default=None)
 
-    parser.add_argument("--save", dest="save", default=True, action="store_true")
-    parser.add_argument("--nosave", dest="save", default=True, action="store_false")
+    parser.add_argument("--save_top_n_val_models", type=int, default=1)
```
Comment (AntreasAntoniou) [inline train.py:66]: Why are there two separate arguments that seem to do the same thing here?

File: utils/metric_tracking.py [hunk 2]
```diff
@@ -145,8 +134,16 @@ def collect_per_epoch(self):
 
         return epoch_metrics
 
-    def get_best_epoch_for_metric(self, metric_name, evaluation_metric=np.argmax):
-        return evaluation_metric(self.collect_per_epoch()[metric_name])
+    def get_best_n_epochs_for_metric(self, metric_name, n=1, bigger_is_better=True):
```
Comment (AntreasAntoniou) [inline utils/metric_tracking.py:137]: lol neat variable name

File: train.py [hunk 3]
```diff
@@ -456,32 +456,35 @@ def eval(
...
+                "model": model.state_dict(),
+                "optimizer": optimizer.state_dict(),
+                "scheduler": scheduler.state_dict(),
+            }
 
+            n_best_epochs = metric_tracker_val.get_best_n_epochs_for_metric(
+                "accuracy", n=args.save_top_n_val_models
+            )
+
+            if epoch in n_best_epochs:
```
Comment (AntreasAntoniou) [inline train.py:472]: Do we really need to keep the 'rank' as part of the naming of the top N models? I can see why it's useful, but it we ommited that, we could just save a model if it's in the top 5, and simply delete the previous worst model. That would save on disk writes. What do you think? It shouldn't matter as it happens once per epoch, though. Having the literal rank in the name makes it a lot easier after the fact to find the model in a folder. hmmm..
What if we have a smarter version of what you are doing? Like, only store the new model that is in the top K, and rename the others to reflect the right rank? This must be a lot cheaper than saving them all again?

Comment (AntreasAntoniou) [reply train.py:472]: You could literally keep a dict of ranks mapped to paths, and if the ranks change simply rename those that remain, and remove the one that doesn't, only saving the new one.

File: train.py (hunk 1 above)
Comment (jack-willturner) [reply train.py:66]: previously the train script had `--save` and also  `--nosave` - I presumed this was you, so replaced it with a single arg.

File: train.py [hunk 4]
```diff
@@ -59,11 +59,11 @@ def get_base_argument_parser():
     # logging
     parser.add_argument("--experiment_name", type=str, default="dev")
     parser.add_argument("--logs_path", type=str, default="log")
+    parser.add_argument("--save_k_top_val_models", type=int, default=1)
```
Comment (AntreasAntoniou) [inline train.py:62]: I am referring to this second argument

File: train.py (hunk 1 above)
Comment (AntreasAntoniou) [reply train.py:66]: Also that was not me. It was there originally.

File: train.py (hunk 4 above)
Comment (AntreasAntoniou) [reply train.py:62]: If you enter review mode you can see that line 62 and 66 have an argument that seems to serve the same purpose

File: train.py (hunk 3 above)
Comment (jack-willturner) [reply train.py:472]: > we could just save a model if it's in the top 5, and simply delete the previous worst model. That would save on disk writes

I'm not sure I understand. Wouldn't we be writing the same number of models either way? 

Unless I did it wrong, this should only save the current epoch weights, we should never re-save old things (+ we don't actually have access to the old weights, so we don't have anything to save there anyway)

I don't love the idea of having code which renames old checkpoints.

Comment (AntreasAntoniou) [reply train.py:472]: Correct me if I am wrong, but the code seems to iterate the top n epoch idx and then save those models with a specific rank built into their names? If not, how do you handle the fact that the previous top model is now the second in rank?

Comment (jack-willturner) [reply train.py:472]: Ah you're right.

I guess we can't do this without having some renaming going on. I'll clear that up.

File: train.py (hunk 4 above)
Comment (jack-willturner) [reply train.py:62]: Forgot this was there! Will remove

**Comment by AntreasAntoniou** [comment]
@jack-willturner Could we get this reviewed and merged sometime soon? Just wanting to get things moving since there's more updates coming very soon

**Review by jack-willturner — COMMENTED** [review]
This all looks good to me :)

**Review by AntreasAntoniou — APPROVED** [review]
